### PR TITLE
DSDEGP-454 Enable the IntelDeflater by default in all invocations of Picard

### PR DESCRIPTION
**Pull request by danxmoran**
### Description

This mirrors what was done for GATK in broadinstitute/gsa-unstable#1572.

Comwell workflow ID in gotc-dev: 7012caa5-ea19-4f99-82f3-a1f2ca200ad6

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [x] Added or modified tests to cover changes and any new functionality
- [x] Edited the README / documentation (if applicable)
- [x] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -15,3 +15,4 @@ report
 jacoco.data
 .gradle
 build
+*.bai
```
Comment (pshapiro4broad) [inline .gitignore]: Why is this necessary? Did a test create a .bai file and not delete it?

File: src/main/java/picard/cmdline/CommandLineProgram.java [hunk 2]
```diff
@@ -102,6 +101,12 @@
 
     @Option(doc="Google Genomics API client_secrets.json file path.", common = true)
     public String GA4GH_CLIENT_SECRETS="client_secrets.json";
+
+    @Option(shortName = "jdk_deflater", doc = "Use the JDK Deflater instead of the IntelDeflater for writing BAMs", common = true)
```
Comment (pshapiro4broad) [inline src/main/java/picard/cmdline/CommandLineProgram.java]: I think it would be better to use the same option name as GATK here, e.g. `use_jdk_deflater`, `use_jdk_inflater`.

File: src/main/java/picard/cmdline/CommandLineProgram.java [hunk 3]
```diff
@@ -190,12 +204,14 @@ public int instanceMain(final String[] argv) {
...
-                                       " on " + System.getProperty("os.name") + " " + System.getProperty("os.version") +
-                                       " " + System.getProperty("os.arch") + "; " + System.getProperty("java.vm.name") +
-                                       " " + System.getProperty("java.runtime.version") +
-                                       "; Picard version: " + commandLineParser.getVersion());
+                final String msg = String.format(
+                    "[%s] Executing as %s@%s on %s %s %s; %s %s; Deflater: %s; Inflater: %s; Picard version: %s",
+                    new Date(), System.getProperty("user.name"), InetAddress.getLocalHost().getHostName(),
+                    System.getProperty("os.name"), System.getProperty("os.version"), System.getProperty("os.arch"),
+                    System.getProperty("java.vm.name"), System.getProperty("java.runtime.version"),
+                    USE_JDK_DEFLATER ? "JdkDeflater" : "IntelDeflater", USE_JDK_INFLATER ? "JdkInflater" : "IntelInflater",
```
Comment (pshapiro4broad) [inline src/main/java/picard/cmdline/CommandLineProgram.java]: If the prefix text is `Deflater` then I think the value text could be just `Intel` or `JDK`.

File: src/main/java/picard/cmdline/CommandLineProgram.java [hunk 4]
```diff
@@ -190,12 +204,14 @@ public int instanceMain(final String[] argv) {
 
             // Output a one liner about who/where and what software/os we're running on
             try {
-            System.err.println("[" + new Date() + "] Executing as " +
-                                       System.getProperty("user.name") + "@" + InetAddress.getLocalHost().getHostName() +
-                                       " on " + System.getProperty("os.name") + " " + System.getProperty("os.version") +
-                                       " " + System.getProperty("os.arch") + "; " + System.getProperty("java.vm.name") +
-                                       " " + System.getProperty("java.runtime.version") +
-                                       "; Picard version: " + commandLineParser.getVersion());
+                final String msg = String.format(
```
Comment (pshapiro4broad) [inline src/main/java/picard/cmdline/CommandLineProgram.java]: Not sure why you changed this to use a variable and `String.format()` instead of using `+`? Do you find this form easier to read?

File: src/main/java/picard/cmdline/CommandLineProgram.java [hunk 5]
```diff
@@ -32,10 +34,7 @@
 import htsjdk.samtools.metrics.MetricBase;
 import htsjdk.samtools.metrics.MetricsFile;
 import htsjdk.samtools.metrics.StringHeader;
-import htsjdk.samtools.util.BlockCompressedOutputStream;
-import htsjdk.samtools.util.BlockCompressedStreamConstants;
-import htsjdk.samtools.util.IOUtil;
-import htsjdk.samtools.util.Log;
+import htsjdk.samtools.util.*;
```
Comment (pshapiro4broad) [inline src/main/java/picard/cmdline/CommandLineProgram.java]: I don't know if we have a policy about this but I think we want to avoid `*` imports in our java code. There is an IDE setting to control this.

File: src/main/java/picard/cmdline/CommandLineProgram.java [hunk 6]
```diff
@@ -86,7 +85,7 @@
     public ValidationStringency VALIDATION_STRINGENCY = ValidationStringency.DEFAULT_STRINGENCY;
 
     @Option(doc = "Compression level for all compressed files created (e.g. BAM and GELI).", common=true)
-    public int COMPRESSION_LEVEL = BlockCompressedStreamConstants.DEFAULT_COMPRESSION_LEVEL;
+    public int COMPRESSION_LEVEL = 1; // The default IntelDeflater is slow at levels > 1
```
Comment (pshapiro4broad) [inline src/main/java/picard/cmdline/CommandLineProgram.java]: IMO end of line comments make code harder to read and edit. In most cases, like this, it can just as easily be a comment on the line before, instead of at the end of the line.

File: .gitignore (hunk 1 above)
Comment (danxmoran) [inline .gitignore]: Yeah, a couple popped out when I ran tests. Should I move this to my global ignore file instead?

File: src/main/java/picard/cmdline/CommandLineProgram.java (hunk 4 above)
Comment (danxmoran) [inline src/main/java/picard/cmdline/CommandLineProgram.java]: I do find this one easier to read. It's a bit harder to figure out which values plug into which slots in the log, but it's easier to see the general structure of the message when it's all on one line at the top.

**Comment by droazen** [comment]
@erniebrau and/or @gspowley could one of you have an initial look at this?

File: src/main/java/picard/cmdline/CommandLineProgram.java [hunk 7]
```diff
@@ -86,7 +88,8 @@
     public ValidationStringency VALIDATION_STRINGENCY = ValidationStringency.DEFAULT_STRINGENCY;
 
     @Option(doc = "Compression level for all compressed files created (e.g. BAM and GELI).", common=true)
-    public int COMPRESSION_LEVEL = BlockCompressedStreamConstants.DEFAULT_COMPRESSION_LEVEL;
+    // The default IntelDeflater is slow at levels > 1
+    public int COMPRESSION_LEVEL = 1;
```
Comment (droazen) [inline src/main/java/picard/cmdline/CommandLineProgram.java]: This comment is not accurate -- the Intel Deflater is faster than the JDK deflater by 15-45% on all compression levels:

image

The reason why we want to compress at level 1 is because reading from a level 1 compressed file is ~3x faster than reading from a level 5 compressed file (the former default), and only slightly larger:

image

File: src/main/java/picard/cmdline/CommandLineProgram.java [hunk 8]
```diff
@@ -102,6 +105,12 @@
 
     @Option(doc="Google Genomics API client_secrets.json file path.", common = true)
     public String GA4GH_CLIENT_SECRETS="client_secrets.json";
+
+    @Option(shortName = "use_jdk_deflater", doc = "Use the JDK Deflater instead of the IntelDeflater for writing BAMs", common = true)
```
Comment (droazen) [inline src/main/java/picard/cmdline/CommandLineProgram.java]: This controls more than just writing BAMs -- it controls anything written via `BlockCompressedOutputStream` (applies to the second argument below as well)

File: src/main/java/picard/cmdline/CommandLineProgram.java [hunk 9]
```diff
@@ -190,12 +208,14 @@ public int instanceMain(final String[] argv) {
...
-                                       " on " + System.getProperty("os.name") + " " + System.getProperty("os.version") +
-                                       " " + System.getProperty("os.arch") + "; " + System.getProperty("java.vm.name") +
-                                       " " + System.getProperty("java.runtime.version") +
-                                       "; Picard version: " + commandLineParser.getVersion());
+                final String msg = String.format(
+                    "[%s] Executing as %s@%s on %s %s %s; %s %s; Deflater: %s; Inflater: %s; Picard version: %s",
+                    new Date(), System.getProperty("user.name"), InetAddress.getLocalHost().getHostName(),
+                    System.getProperty("os.name"), System.getProperty("os.version"), System.getProperty("os.arch"),
+                    System.getProperty("java.vm.name"), System.getProperty("java.runtime.version"),
+                    USE_JDK_DEFLATER ? "Jdk" : "Intel", USE_JDK_INFLATER ? "Jdk" : "Intel",
```
Comment (droazen) [inline src/main/java/picard/cmdline/CommandLineProgram.java]: You need to test whether the Intel Deflater/Inflater were actually successfully loaded, and use that result to inform the user which deflater/inflater will be used. What you want to do here is:

```
        final boolean usingIntelDeflater = (BlockCompressedOutputStream.getDefaultDeflaterFactory() instanceof IntelDeflaterFactory && ((IntelDeflaterFactory)BlockCompressedOutputStream.getDefaultDeflaterFactory()).usingIntelDeflater());
        final boolean usingIntelInflater = (BlockGunzipper.getDefaultInflaterFactory() instanceof IntelInflaterFactory && ((IntelInflaterFactory)BlockGunzipper.getDefaultInflaterFactory()).usingIntelInflater());
```

File: src/main/java/picard/cmdline/CommandLineProgram.java [hunk 10]
```diff
@@ -190,12 +208,14 @@ public int instanceMain(final String[] argv) {
...
+                    commandLineParser.getVersion());
+                System.err.println(msg);
             }
             catch (Exception e) { /* Unpossible! */ }
         }
```
Comment (droazen) [inline src/main/java/picard/cmdline/CommandLineProgram.java]: This branch needs a unit test. Recommend adding a test proving that the Intel Inflater and Deflater can be loaded successfully.

File: src/main/java/picard/cmdline/CommandLineProgram.java (hunk 5 above)
Comment (droazen) [inline src/main/java/picard/cmdline/CommandLineProgram.java]: Collapsing imports after a set threshold is standard practice.

File: build.gradle [hunk 11]
```diff
@@ -50,6 +50,7 @@ jacoco {
 final htsjdkVersion = System.getProperty('htsjdk.version', '2.10.1')
 
 dependencies {
+    compile 'com.intel.gkl:gkl:0.5.2'
```
Comment (danxmoran) [inline build.gradle]: screen shot 2017-07-05 at 9 43 04 am
Confirmed! Is this something that would be good to assert on in our jar-building Gradle tasks?

File: src/test/java/picard/IntelInflaterDeflaterLoadTest.java [hunk 12]
```diff
@@ -0,0 +1,37 @@
+package picard;
```
Comment (danxmoran) [inline src/test/java/picard/IntelInflaterDeflaterLoadTest.java]: Here's a test for the successful loading of the Intel inflater / deflater, modeled off of the [GATK tests](https://github.com/broadinstitute/gatk/blob/master/src/test/java/org/broadinstitute/hellbender/IntelInflaterDeflaterIntegrationTest.java).

File: src/test/java/picard/analysis/CollectMultipleMetricsTest.java [hunk 13]
```diff
@@ -335,8 +335,10 @@ void setupBuilder() throws IOException {
         final String flowCellBarcode = "TESTBARCODE";
 
         tempSamFile = File.createTempFile("CollectGcBias", ".bam", TEST_DIR);
+        final File tempSamIndex = new File(tempSamFile.getAbsolutePath().replace("bam", "bai"));
```
Comment (danxmoran) [inline src/test/java/picard/analysis/CollectMultipleMetricsTest.java]: This test would successfully clean up after itself when run alone with `-Dtest.single=CollectMultipleMetricsTest` but fail to clean up when run with all tests in the `analysis` package (or all the tests in Picard). Explicitly marking the index file for deletion fixes the problem, but I'm not sure what the root issue is.

File: src/test/java/picard/analysis/CollectWgsMetricsTest.java [hunk 14]
```diff
@@ -136,8 +136,10 @@ void setupBuilder() throws IOException {
 
         //Create Sam Files
         tempSamFile = File.createTempFile("CollectWgsMetrics", ".bam", TEST_DIR);
+        final File tempSamIndex = new File(tempSamFile.getAbsolutePath().replace("bam", "bai"));
```
Comment (danxmoran) [inline src/test/java/picard/analysis/CollectWgsMetricsTest.java]: Same as in `CollectMultipleMetricsTest`.

File: src/test/java/picard/sam/FilterSamReadsTest.java [hunk 15]
```diff
@@ -95,7 +95,7 @@ public void testPairedIntervalFilter(final String intervalFilename, final int ex
         // Build a sam file for testing
         final File inputSam = File.createTempFile("testSam", ".sam", TEST_DIR);
         inputSam.deleteOnExit();
-        final File sortedSamIdx = new File(TEST_DIR, inputSam.getName() + ".idx");
+        final File sortedSamIdx = new File(TEST_DIR, inputSam.getName() + ".bai");
```
Comment (danxmoran) [inline src/test/java/picard/sam/FilterSamReadsTest.java]: This test reliably failed to clean up after itself because of the wrong extension here.

File: src/main/java/picard/cmdline/CommandLineProgram.java (hunk 10 above)
Comment (danxmoran) [inline src/main/java/picard/cmdline/CommandLineProgram.java]: Added below.

**Comment by danxmoran** [comment]
Ok @droazen, comments addressed!

File: src/main/java/picard/cmdline/CommandLineProgram.java [hunk 16]
```diff
@@ -102,6 +106,12 @@
 
     @Option(doc="Google Genomics API client_secrets.json file path.", common = true)
     public String GA4GH_CLIENT_SECRETS="client_secrets.json";
+
+    @Option(shortName = "use_jdk_deflater", doc = "Use the JDK Deflater instead of the Intel Deflater for writing with BlockCompressedOutputStreams", common = true)
+    public Boolean USE_JDK_DEFLATER = false;
+
+    @Option(shortName = "use_jdk_inflater", doc = "Use the JDK Inflater instead of the Intel Inflater for reading with BlockGunzippers", common = true)
+    public Boolean USE_JDK_INFLATER = false;
```
Comment (droazen) [inline src/main/java/picard/cmdline/CommandLineProgram.java]: These argument docs should be targeted at end users rather than developers, and should not mention class names. Eg.,

```
"Use the JDK Deflater instead of the Intel Deflater for writing compressed output"
"Use the JDK Inflater instead of the Intel Inflater for reading compressed input"
```

File: src/main/java/picard/cmdline/CommandLineProgram.java [hunk 17]
```diff
@@ -190,12 +209,18 @@ public int instanceMain(final String[] argv) {
...
+                final boolean usingIntelDeflater = (BlockCompressedOutputStream.getDefaultDeflaterFactory() instanceof IntelDeflaterFactory &&
+                        ((IntelDeflaterFactory)BlockCompressedOutputStream.getDefaultDeflaterFactory()).usingIntelDeflater());
+                final boolean usingIntelInflater = (BlockGunzipper.getDefaultInflaterFactory() instanceof IntelInflaterFactory &&
+                        ((IntelInflaterFactory)BlockGunzipper.getDefaultInflaterFactory()).usingIntelInflater());
+                final String msg = String.format(
+                    "[%s] Executing as %s@%s on %s %s %s; %s %s; Deflater: %s; Inflater: %s; Picard version: %s",
+                    new Date(), System.getProperty("user.name"), InetAddress.getLocalHost().getHostName(),
+                    System.getProperty("os.name"), System.getProperty("os.version"), System.getProperty("os.arch"),
+                    System.getProperty("java.vm.name"), System.getProperty("java.runtime.version"),
+                    usingIntelDeflater ? "Jdk" : "Intel", usingIntelInflater ? "Jdk" : "Intel",
```
Comment (droazen) [inline src/main/java/picard/cmdline/CommandLineProgram.java]: These expressions giving the deflater/inflater being used are backwards. Should be:

```
usingIntelDeflater ? "Intel" : "Jdk"
```

and

```
usingIntelInflater ? "Intel" : "Jdk"
```

File: src/test/java/picard/sam/FilterSamReadsTest.java (hunk 15 above)
Comment (droazen) [inline src/test/java/picard/sam/FilterSamReadsTest.java]: A sam file can't have a `.bai` index though?

File: src/main/java/picard/cmdline/CommandLineProgram.java (hunk 17 above)
Comment (danxmoran) [inline src/main/java/picard/cmdline/CommandLineProgram.java]: D'oh

File: src/test/java/picard/sam/FilterSamReadsTest.java (hunk 15 above)
Comment (danxmoran) [inline src/test/java/picard/sam/FilterSamReadsTest.java]: I can't comment on [the line](https://github.com/broadinstitute/picard/pull/843/files#diff-9c09d6174fd9f608424a44e8e09ac593R102) directly, but could it be because the test is using `makeBAMWriter` instead of `makeSAMWriter` below?

File: src/test/java/picard/sam/FilterSamReadsTest.java [hunk 18]
```diff
@@ -99,7 +99,7 @@ public void testPairedIntervalFilter(final String intervalFilename, final int ex
         sortedSamIdx.deleteOnExit();
 
         final SAMFileWriter writer = new SAMFileWriterFactory()
-                .setCreateIndex(true).makeBAMWriter(builder.getHeader(), false, inputSam);
+                .setCreateIndex(true).makeSAMWriter(builder.getHeader(), false, inputSam);
```
Comment (danxmoran) [inline src/test/java/picard/sam/FilterSamReadsTest.java]: This change also stops the test from leaving index files lying around. It also seems to make more sense given the `inputSam` name and extension, but I honestly don't know enough about BAM vs SAM to know if this is the right move. The test passes either way.

Comment (droazen) [inline src/test/java/picard/sam/FilterSamReadsTest.java]: Looks like the right move to me -- the test case explicitly claims to "Build a sam file for testing" (rather than a BAM file)

File: src/test/java/picard/analysis/CollectMultipleMetricsTest.java (hunk 13 above)
Comment (sooheelee) [inline src/test/java/picard/analysis/CollectMultipleMetricsTest.java]: Is the index a valid index? If you run ValidateSamFile on the bam/bai pair, do the files validate?
